torch/csrc/jit/codegen/cuda/partition.cpp

```diff
@@ -71,6 +71,222 @@ bool hasReductionOperation(const Node* node) {
   return false;
 }
 
+// utility function to check if the node implies broadcast on a given shape (
+// assumed to be shape of an input tensor)
+// limitations:
+//   1. we rely on shape information to judge this. so we would require output
+//      shape to be available;
+//   2. we basically compares given shape to the shape of the only output of
+//      the node and return true if it implies broadcast from the former to the
+//      latter.
+bool maybeBroadcastOnShape(
+    const Node* n,
+    const std::vector<c10::optional<int64_t>>& shape) {
+  TORCH_INTERNAL_ASSERT(
+      n->outputs().size() == 1,
+      "not expecting multiple outputs from a node, graph partitioning logic needs to be updated");
+  // assumes that if output is not a tensor type, it's not broadcasting
+  if (auto out_type = n->output(0)->type()->cast<TensorType>()) {
+    if (out_type->dim()) {
+      if (out_type->dim().value() < shape.size()) {
+        // no broadcast for reduction operation;
+        return false;
+      } else if (out_type->dim().value() > shape.size()) {
+        // increased rank means there is reduction;
+        return true;
+      } else {
+        // same rank, we need to iterate through sizes and check if size-1
+        // exists in input `shape`
+        for (const auto& opt_size : shape) {
+          // TODO: not sure if we need to check for output size != 1, since we
+          // are currently marking all size-1 dimension as broadcast in codegen.
```

Comment:
Yes you should. We mark it as broadcast, but something is only "truly" a broadcast when it gets broadcasted. Broadcast op is almost like a "maybe broadcast".

Reply:
Doesn't have to be in this PR, can leave this TODO for later.

Reply:
Thanks, I'd like to keep it TODO for now.

I'm only worried that it would still trigger propagation failure even if it's not a real broadcast.